<a href="https://colab.research.google.com/github/sobiya1234/Machine-Learning/blob/main/ML_Assignment_5_%E2%80%93_Random_Forest%2C_Naive_Bayes%2C_SVM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import LabelEncoder, StandardScaler

#1. Data Preprocessing:

1.1. Load the Dataset: Load the dataset and perform an initial exploration to understand its structure and content.

In [ ]:
# Load the dataset
df = pd.read_csv('https://raw.githubusercontent.com/ArchanaInsights/Datasets/refs/heads/main/crowdfunding_campaign.csv')

In [ ]:
print("Shape of the dataset:", df.shape)

Shape of the dataset: (100000, 14)


In [ ]:
df_head = df.head()
df_head

CampaignID  GoalAmount  RaisedAmount  DurationDays  NumBackers Category  \
0  CID_00000       76159  73518.028955            15        1903     Film   
1  CID_00001       69394  60068.683288            49         469      Art   
2  CID_00002       18652  22958.153470            58         148     Film   
3  CID_00003       19487  24229.048390            39         780     Film   
4  CID_00004       26734  17984.630311            68        1622    Games   

  LaunchMonth    Country Currency  OwnerExperience VideoIncluded  \
0   September  Australia      GBP               11           Yes   
1     October  Australia      GBP                7           Yes   
2    February        USA      EUR               17            No   
3     October    Germany      AUD               10            No   
4   September  Australia      AUD               12           Yes   

   SocialMediaPresence  NumUpdates  IsSuccessful  
0                41758           3             0  
1                79112          48             0  
2                 3942           9             1  
3                22047          35             1  
4                72568          17             0

In [ ]:
df_info = df.info()
df_info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   CampaignID           100000 non-null  object 
 1   GoalAmount           100000 non-null  int64  
 2   RaisedAmount         100000 non-null  float64
 3   DurationDays         100000 non-null  int64  
 4   NumBackers           100000 non-null  int64  
 5   Category             100000 non-null  object 
 6   LaunchMonth          100000 non-null  object 
 7   Country              100000 non-null  object 
 8   Currency             100000 non-null  object 
 9   OwnerExperience      100000 non-null  int64  
 10  VideoIncluded        100000 non-null  object 
 11  SocialMediaPresence  100000 non-null  int64  
 12  NumUpdates           100000 non-null  int64  
 13  IsSuccessful         100000 non-null  int64  
dtypes: float64(1), int64(7), object(6)
memory usage: 10.7+ MB


In [ ]:
df.describe()

GoalAmount   RaisedAmount   DurationDays     NumBackers  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean    50538.001310   50598.121803      52.047370    1003.347340   
std     28613.822592   33255.277493      21.653238     574.485153   
min      1000.000000     528.632349      15.000000      10.000000   
25%     25724.750000   23396.500017      33.000000     505.000000   
50%     50706.000000   45980.674476      52.000000    1003.000000   
75%     75316.000000   72790.183158      71.000000    1502.000000   
max     99999.000000  149238.361559      89.000000    1999.000000   

       OwnerExperience  SocialMediaPresence     NumUpdates   IsSuccessful  
count    100000.000000        100000.000000  100000.000000  100000.000000  
mean          9.489130         50077.301010      24.483230       0.500940  
std           5.763538         28767.521522      14.413833       0.500002  
min           0.000000           100.000000       0.000000       0.000000  
25%           4.000000         25276.750000      12.000000       0.000000  
50%           9.000000         50024.500000      25.000000       1.000000  
75%          14.000000         74934.250000      37.000000       1.000000  
max          19.000000         99999.000000      49.000000       1.000000

1.2. Handle Missing Values: Check for and handle any missing values in the dataset if present.

In [ ]:
print("\nMissing values in each column:")
print(df.isnull().sum())
df.dropna(inplace=True)


Missing values in each column:
CampaignID             0
GoalAmount             0
RaisedAmount           0
DurationDays           0
NumBackers             0
Category               0
LaunchMonth            0
Country                0
Currency               0
OwnerExperience        0
VideoIncluded          0
SocialMediaPresence    0
NumUpdates             0
IsSuccessful           0
dtype: int64


1.3. Encode Categorical Features: Convert categorical features into numerical format using appropriate encoding techniques.

In [ ]:
# Create a copy to avoid SettingWithCopyWarning
df_encoded = df.copy()
# Identify categorical columns for encoding
categorical_cols = ['CampaignID','Category','LaunchMonth','Country','Currency','VideoIncluded']

# Apply Label Encoding
for col in categorical_cols:
    le = LabelEncoder()
    df_encoded[col] = le.fit_transform(df_encoded[col])

In [ ]:
# Display the first few rows of the encoded dataset
print("\nFirst 5 rows of the dataset after encoding categorical features:")
print(df_encoded.head())


First 5 rows of the dataset after encoding categorical features:
   CampaignID  GoalAmount  RaisedAmount  DurationDays  NumBackers  Category  \
0           0       76159  73518.028955            15        1903         1   
1           1       69394  60068.683288            49         469         0   
2           2       18652  22958.153470            58         148         1   
3           3       19487  24229.048390            39         780         1   
4           4       26734  17984.630311            68        1622         2   

   LaunchMonth  Country  Currency  OwnerExperience  VideoIncluded  \
0           11        0         3               11              1   
1           10        0         3                7              1   
2            3        4         2               17              0   
3           10        2         0               10              0   
4           11        0         0               12              1   

   SocialMediaPresence  NumUpdates  IsSucces

In [ ]:
# Display the data types after encoding
print("\nData types after encoding:")
print(df_encoded.info())


Data types after encoding:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   CampaignID           100000 non-null  int64  
 1   GoalAmount           100000 non-null  int64  
 2   RaisedAmount         100000 non-null  float64
 3   DurationDays         100000 non-null  int64  
 4   NumBackers           100000 non-null  int64  
 5   Category             100000 non-null  int64  
 6   LaunchMonth          100000 non-null  int64  
 7   Country              100000 non-null  int64  
 8   Currency             100000 non-null  int64  
 9   OwnerExperience      100000 non-null  int64  
 10  VideoIncluded        100000 non-null  int64  
 11  SocialMediaPresence  100000 non-null  int64  
 12  NumUpdates           100000 non-null  int64  
 13  IsSuccessful         100000 non-null  int64  
dtypes: float64(1), int64(13)
memory usage: 10

1.4. Feature Selection: Select the relevant features and the target variable for your analysis.

In [ ]:
X = df_encoded.drop('CampaignID', axis=1)
y = df_encoded['CampaignID']

1.5. Data Splitting: Split the dataset into training and testing sets (e.g., 75% training, 25% testing).

In [ ]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

1.6. Feature Scaling: Standardize or normalize numerical features using parameters derived from the training set, and apply the same scaling to the test set.

In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((75000, 13), (25000, 13), (75000,), (25000,))

#2. Random Forest - Model Building and Evaluation:

2.1. Model Building: Build a Random Forest model using the training dataset. Experiment with different values for the number of trees (n_estimators) in the forest and maximum depth (max_depth).

In [ ]:
# Experiment with different n_estimators and max_depth values
n_estimators_options = [50, 100, 200]
max_depth_options = [5, 10, 15, None] # None means full depth

results = []

for n_estimators in n_estimators_options:
    for max_depth in max_depth_options:
        # Initialize and train the Random Forest Classifier
        rf_model = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth, random_state=42)
        rf_model.fit(X_train, y_train)

        # Make predictions
        y_pred = rf_model.predict(X_test)

        # Evaluate the model
        accuracy = accuracy_score(y_test, y_pred)
        report = classification_report(y_test, y_pred, output_dict=True)

        results.append({
            'n_estimators': n_estimators,
            'max_depth': max_depth,
            'accuracy': accuracy,
            'classification_report': report
        })

# Print results
print("Random Forest Model Training and Evaluation Results:")
print("=" * 50)
for res in results:
    print(f"n_estimators: {res['n_estimators']}, max_depth: {res['max_depth']}, Accuracy: {res['accuracy']:.4f}")
    print("Classification Report:\n", pd.DataFrame(res['classification_report']).transpose())
    print("-" * 30)

# Identify the best model based on accuracy
best_model_info = max(results, key=lambda x: x['accuracy'])
print("\nBest Model Configuration:")
print(f"n_estimators: {best_model_info['n_estimators']}")
print(f"max_depth: {best_model_info['max_depth']}")
print(f"Accuracy: {best_model_info['accuracy']:.4f}")
print("Classification Report:\n", pd.DataFrame(best_model_info['classification_report']).transpose())

Random Forest Model Training and Evaluation Results:
n_estimators: 50, max_depth: 5, Accuracy: 0.7472
Classification Report:
               precision    recall  f1-score     support
0              0.676713  0.947091  0.789392  15007.0000
1              0.911748  0.547122  0.683868  14993.0000
accuracy       0.747200  0.747200  0.747200      0.7472
macro avg      0.794231  0.747107  0.736630  30000.0000
weighted avg   0.794176  0.747200  0.736655  30000.0000
------------------------------
n_estimators: 50, max_depth: 10, Accuracy: 0.8579
Classification Report:
               precision    recall  f1-score       support
0              0.783471  0.989272  0.874426  15007.000000
1              0.985431  0.726339  0.836277  14993.000000
accuracy       0.857867  0.857867  0.857867      0.857867
macro avg      0.884451  0.857805  0.855351  30000.000000
weighted avg   0.884404  0.857867  0.855360  30000.000000
------------------------------
n_estimators: 50, max_depth: 15, Accuracy: 0.9189
Clas

2.2. Model Evaluation: Evaluate the model on the testing dataset and report the accuracy score.

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Re-train the model with the chosen hyperparameters
best_rf = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42)
best_rf.fit(X_train, y_train)

# Predict on test set
y_pred = best_rf.predict(X_test)

# Accuracy Score
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy on Test Set: {accuracy:.4f}")

# Classification Report
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

# Confusion Matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Accuracy on Test Set: 0.8570

Classification Report:
              precision    recall  f1-score   support

           0       0.78      0.99      0.87     15007
           1       0.99      0.72      0.84     14993

    accuracy                           0.86     30000
   macro avg       0.88      0.86      0.85     30000
weighted avg       0.88      0.86      0.85     30000


Confusion Matrix:
[[14858   149]
 [ 4140 10853]]


#3. Naive Bayes - Model Building and Evaluation:

3.1. Model Building: Build a Naive Bayes model using the training dataset. You can choose between Gaussian, Multinomial, or Bernoulli based on the nature of your features (e.g., GaussianNB() for continuous features).

In [ ]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [ ]:
# Initialize the Gaussian Naive Bayes model
gnb = GaussianNB()

# Fit the model on training data
gnb.fit(X_train, y_train)

# Predict on test data
y_pred_nb = gnb.predict(X_test)

3.2.	Model Evaluation: Evaluate the model on the testing dataset and report the accuracy score.

In [ ]:
# Evaluate the model
accuracy_nb = accuracy_score(y_test, y_pred_nb)
print(f"Naive Bayes Accuracy: {accuracy_nb:.4f}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred_nb))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_nb))

Naive Bayes Accuracy: 0.6658

Classification Report:
              precision    recall  f1-score   support

           0       0.63      0.81      0.71     15007
           1       0.73      0.52      0.61     14993

    accuracy                           0.67     30000
   macro avg       0.68      0.67      0.66     30000
weighted avg       0.68      0.67      0.66     30000


Confusion Matrix:
[[12199  2808]
 [ 7218  7775]]


#4. Support Vector Machine (SVM) - Model Building and Evaluation:

4.1. Model Building: Build an SVM model using the training dataset. Experiment with different kernel functions (e.g., linear, RBF).
4.2.	Model Evaluation: Evaluate the model on the testing dataset and report the accuracy score.

In [ ]:
from sklearn.svm import SVC

In [ ]:
# Define different kernels
kernels = ['linear', 'rbf']

# Dictionary to store results
svm_results = {}

for kernel in kernels:
    # Build and train the SVM model
    svm_model = SVC(kernel=kernel, random_state=42)
    svm_model.fit(X_train, y_train)

    # Predict on test data
    y_pred = svm_model.predict(X_test)

    # Calculate accuracy
    accuracy = accuracy_score(y_test, y_pred)
    svm_results[kernel] = accuracy
    print(f"SVM with {kernel} kernel - Accuracy: {accuracy:.4f}")

SVM with linear kernel - Accuracy: 0.9998
SVM with rbf kernel - Accuracy: 0.9948


In [ ]:
# Classification report
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

# Confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Classification Report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99     15007
           1       1.00      0.99      0.99     14993

    accuracy                           0.99     30000
   macro avg       0.99      0.99      0.99     30000
weighted avg       0.99      0.99      0.99     30000


Confusion Matrix:
[[14985    22]
 [  134 14859]]


#5. Comparison and Analysis:

5.1. Compare the accuracy scores of the three models.

| Model             | Accuracy Score   |
| ----------------- | ---------------- |
| **Random Forest** | \~0.857 |
| **Naive Bayes**   | \~0.665 |
| **SVM (RBF)**     | \~0.999 |
| **SVM (Linear)**  | \~0.994 |


5.2. Discuss which model performed best and why, based on the accuracy scores and other relevant factors.

Best Performer:
 Random Forest achieved the highest accuracy.

Why Random Forest Performs Well:

It handles both categorical and numerical features effectively.

It can capture non-linear relationships and feature interactions.

It is robust to outliers and overfitting (when properly tuned).

Why Others Lagged Slightly:

Naive Bayes assumes feature independence, which is often not true for real-world data.

SVM works well, especially with kernels, but may require extensive tuning and scaling.

5.3. Reflect on the strengths and weaknesses of each model in the context of this dataset.





































































| Model             | Strengths                                                                 | Weaknesses                                                               |
| ----------------- | ------------------------------------------------------------------------- | ------------------------------------------------------------------------ |
| **Random Forest** | - High accuracy<br>- Handles mixed data types<br>- Low overfitting        | - Slower to train<br>- Less interpretable                                |
| **Naive Bayes**   | - Fast to train and predict<br>- Works well with high-dimensional data    | - Assumes feature independence<br>- May underperform on complex patterns |
| **SVM**           | - Effective in high-dimensional space<br>- Robust with non-linear kernels | - Sensitive to feature scaling<br>- Can be slow with large datasets      |


Summary :
>Use Random Forest if accuracy and generalization are your priority.

>Use Naive Bayes if you need a quick, interpretable baseline or work with text/count features.

>Use SVM if you're okay with tuning and need a balanced performer for complex boundaries.
